sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/PlanTest.scala

```diff
@@ -51,6 +51,15 @@ trait CodegenInterpretedPlanTest extends PlanTest {
       super.test(testName + " (interpreted path)", testTags: _*)(testFun)(pos)
     }
   }
+
+  protected def testFallback(
+      testName: String,
+      testTags: Tag*)(testFun: => Any)(implicit pos: source.Position): Unit = {
+    val codegenMode = CodegenObjectFactoryMode.FALLBACK.toString
+    withSQLConf(SQLConf.CODEGEN_FACTORY_MODE.key -> codegenMode) {
+      super.test(testName, testTags: _*)(testFun)(pos)
+    }
```

Comment:
Maybe we should run twice, one for `FALLBACK` and one for `NO_CODEGEN`.

`FALLBACK` will run codegen path first and fallback to interpreted mode if codegen fails. So interpreted mode may not run if codegen mode successes.

Reply:
That sounds like another use case. Technically, this `testFallback` is designed to the **default** mode of this configuration (which is the default of Apache Spark runtime in non-testing mode) instead of a specific mode `interpreted mode` or `codeine`.

Reply:
Maybe, do you want to add `testCodegenFail`?

Reply:
Hmm. At the second thought, I understand what was your suggestion. So, your suggestion is to check the occurrence of `fallback`, right? Then, it makes sense.

Reply:
I'll update my PR in this afternoon~

Reply:
`test` runs the test code twice, one for codegen and one for interpreted. It is for test coverage.

If `testFallback` only runs for fallback mode, then it might only runs codegen and skips interpreted mode if codegen successes. So the test coverage is less, I think.

Reply:
Hi, @viirya . I tried to revise, but it seems strange to me.

> test runs the test code twice, one for codegen and one for interpreted. It is for test coverage.
> If testFallback only runs for fallback mode, then it might only runs codegen and skips interpreted mode if codegen successes. So the test coverage is less, I think.

`testFallback` is not aiming to replace `test` in [line 41](https://github.com/apache/spark/pull/31764/files#diff-803f946128579a789870dc0bc6f58602db77476f8c04968a72c56efa783c12a1R41). Instead, `testFallback` is added because `PlanTest` bans all derived classes to test `FALLBACK` mode. Before this PR, there is no way to test `FALLBACK` mode.

As we know, `CODEGEN_FACTORY_MODE` has three modes: `FALLBACK`, `CODEGEN_ONLY`, `NO_CODEGEN`.
As we guess in the name, `testFallback`, this test is specifically for `FALLBACK` mode and is added for some test cases which works only at `FALLBACK` mode as @rednaxelafx described.

So, `testFallback` doesn't aim to mean (1) `CODEGEN_ONLY` should fails and (2) `NO_CODEGEN` should passed. If you want to test this (both (1) and (2)), we should make another test function like `testCodegenFailNoCodegenPass`.

Reply:
`FALLBACK` mode is not a special path other than codegen and interpreted path. Under `FALLBACK` mode, Spark runs first codegen path and then interpreted path if codegen fails.

So it sounds weird that some test cases works only at `FALLBACK` mode. Doesn't it just mean the test cases may fail under codegen but can success under interpreted?

Wrapping the test function with `FALLBACK` config, means we may only run the test in codegen path, if codegen path successes. It skips interpreted path in the case.

So, it may unintentionally avoid the test coverage of interpreted path.

Reply:
Hi, @viirya . It seems that you missed @rednaxelafx 's comment, https://github.com/apache/spark/pull/31709#issuecomment-791275810 .
> So it sounds weird that some test cases works only at FALLBACK mode, so we only need to run it with FALLBACK mode. Doesn't it just mean the test cases may fail under codegen but can success under interpreted?